src/util/locking/FileSystemResourceLocker.ts

```diff
@@ -0,0 +1,160 @@
+import { createHash } from 'crypto';
+import { ensureDirSync, pathExists, readdir, rmdir } from 'fs-extra';
+import type { LockOptions, UnlockOptions } from 'proper-lockfile';
+import { lock, unlock } from 'proper-lockfile';
+import type { ResourceIdentifier } from '../../http/representation/ResourceIdentifier';
+import type { Finalizable } from '../../init/final/Finalizable';
+import { getLoggerFor } from '../../logging/LogUtil';
+import { createErrorMessage } from '../errors/ErrorUtil';
+import { InternalServerError } from '../errors/InternalServerError';
+import type { AttemptSettings } from '../LockUtils';
+import { retryFunction } from '../LockUtils';
+import { joinFilePath } from '../PathUtil';
+import type { ResourceLocker } from './ResourceLocker';
+
+type BooleanOrVoid = boolean | void;
+type VoidFunctionOrVoid = (() => Promise<void>) | void;
+
+const defaultLockOptions: LockOptions = {
+  /** Resolve symlinks using realpath, defaults to true (note that if true, the file must exist previously) */
+  realpath: false,
+  /** The number of retries or a [retry](https://www.npmjs.org/package/retry) options object, defaults to 0 */
+  retries: 0,
+};
+
+const defaultUnlockOptions: UnlockOptions = {
+  /** Resolve symlinks using realpath, defaults to true (note that if true, the file must exist previously) */
+  realpath: false,
+};
+
+const attemptDefaults: Required<AttemptSettings> = { retryCount: -1, retryDelay: 50, retryJitter: 30 };
+
+/**
+ * Argument interface of the FileSystemResourceLocker constructor.
+ */
+interface FileSystemResourceLockerArgs {
+  /** The rootPath of the filesystem */
+  rootFilePath?: string;
+  /** The path to the directory where locks will be stored (appended to rootFilePath) */
+  lockDirectory?: string;
+  /** Custom settings concerning retrying locks */
+  attemptSettings?: AttemptSettings;
+}
+
+function isCodedError(err: unknown): err is { code: string } & Error {
+  return typeof err === 'object' && err !== null && 'code' in err;
+}
+
+/**
+ * A resource locker making use of the [proper-lockfile](https://www.npmjs.com/package/proper-lockfile) library.
+ * Note that no locks are kept in memory, thus this is considered thread- and process-safe.
+ *
+ * This **proper-lockfile** library has its own retry mechanism for the operations, since a lock/unlock call will
+ * either resolve successfully or reject immediately with the causing error. The retry function of the library
+ * however will be ignored and replaced by our own LockUtils' {@link retryFunctionUntil} function.
+ *
+ * This allows us to reuse the AttemptSettings semantics similar to the {@link RedisLocker} class.
```

Comment:
```suggestion
```
Would not reference redis locker just because they are similar. Especially since the redis components might at some point move to a separate repository.